rfcs/0097-no-read-store-dir.md

```diff
@@ -0,0 +1,80 @@
+---
+feature: nix-store-perms
+start-date: 2021-07-04
+author: Las Safin
+co-authors:
+shepherd-team: @kevincox @7c6f434c @edolstra
+shepherd-leader: @edolstra
+related-issues:
+---
+
+# Summary
+[summary]: #summary
+
+- NixOS should have a module for configuring the permissions set for `/nix/store` on boot.
+- Nix should not enforce the permissions used for `/nix/store`.
+- The default permissions if the store doesn't exist should be 1735 when the store is made by Nix or the NixOS installer.
+  This means that the nixbld group can't `ls` the directory.
+
+# Motivation
+[motivation]: #motivation
+
+Right now you can't set the permissions for `/nix/store`, since they'll be overwritten
+by Nix anytime you use `nix`.
+
+`chmod g-r /nix/store` is beneficial because the `nixbld` group doesn't actually
+need to read the directory. It only needs to be able to write and "execute" it.
+This, however, should be optional, since the user should be able to configure
+the permissions however they want.
+
+Some users might also want to do things like `chmod o-r /nix/store`, which
+gives you the interesting property that you can not access paths you do not
+already know of.
+Do note that given that all processes can by default read `/proc/cmdline`,
+`/run/current-system`, and many other places which reveal your
+system's closure, making this permission change an insufficient solution for
+security in many cases. This, however, is also entirely optional and is not
+the default in any way.
```

Comment:
I assume that at least some filesystems are vulnerable to timing attacks, probing paths with `stat` to build prefixes of subdirectory paths until the full path is recovered; using the `x` bit to recover `r` for all paths, including ones that aren't running.

Nonetheless, it's good for practical hermeticity.

Reply:
Yes, we hope that people either look at the list and start, like you, name extra leak vectors, or trust «insufficient for security» claim. It would be interesting to have a survey of path leak vectors, but this RFC is shorter than any useful survey like that…

Reply:
I agree. Any list we try to create in this RFC is likely incomplete so it is better to just state that as an assumption and accept it for the rationale here. If someone wants to prove that the list of leaks is empty (and expected to stay that way) they can propose a follow-up RFC to rely on that fact.